README.md
# Natural Mystic Shaders

**Natural Mystic Shaders** is a shader pack for Minecraft Bedrock
aiming to be as realistic as possible.

![Screenshot of a hill](./img/hill.jpg)
![Screenshot of a river](./img/river.jpg)

Features include:

* Shadows
* Non-blocky clouds
* Waves for water and leaves (but not plants)
* Camera-angle dependent transparency and specular light for water
* Improved torch light with flickering effect
* Sunlight and moonlight colors with night time desaturation
* Desaturation under a bad weather condition
* Exposure adjustment, contrast filter, and tone mapping
* Shader-based rendering of the sun and the moon.

## Download

See [releases](https://github.com/depressed-pho/natural-mystic-shaders/releases).

## Release notes

See [NEWS](NEWS.md).

## Supported platforms

* iOS
* Android (untested)

## Tested on

* Minecraft Bedrock 1.16.221, iPad Pro (MLMY2J/A), iOS 12.1 (16B92)

## Changing configuration

Ideally there should be an in-game GUI to dynamically control the
settings of shaders, but that isn't possible at the moment
([#44](https://github.com/depressed-pho/natural-mystic-shaders/issues/44)). To
compensate this shader pack contains a configuration file so users can
edit it directly to modify its behavior.

* Unpack the `.mcpack` file to some directory. It's just a zip archive.
* Open `src/shaders/glsl/natural-mystic-config.h` with your favorite editor.
* Edit the file to change configuration.
* Archive it again, and install it.

For example, you can disable the shader-generated clouds and turn it
into the default by changing the following line:

```glsl
#define ENABLE_FBM_CLOUDS 1
```

To this:

```glsl
//#define ENABLE_FBM_CLOUDS 1
```

## Author

PHO

## License
[CC0](https://creativecommons.org/share-your-work/public-domain/cc0/)
“No Rights Reserved”


NEWS.md
# Release notes

## 1.9.1 -- not released yet

* Improved compatibility with GLSL ES implementations where fragment
  shaders don't support ``highp`` (#86).
* Replace all use of ``TIME`` with ``TOTAL_REAL_WORLD_TIME`` (#87).
* Added a configuration item ``ENABLE_CAMERA_MOVEMENT_MITIGATION``
  which is disabled by default. It mitigates the problem where the
  animation of water and leaves move along with players, but has a
  side effect of causing a discontinuity when a player crosses chunk
  borders (#85).

## 1.9.0 -- 2021-05-09

* Updated the base code from 1.16.0 to 1.16.221.
* Implemented shaders for shader-generated sun and moon (#52). The
  moon shader does a diffuse lighting to render lunar phases.
* Added a configuration item ``ENABLE_SHADER_SUN_MOON`` which is
  enabled by default. Disabling it will revert the sun and the moon to
  the default texture-based rendering.
* The waving animation of plants used to happen even indoors. It is
  now fixed (#85).
* The waving animation of plants used to have a discontinuity at each
  of chunk borders. It is now fixed too (#85).
* The perturbance of water also significantly decreases indoors now.

## 1.8.0 -- 2020-06-25

The 1.16.0 nether upgrade surprisingly didn't break my shaders too
much so this is only a small fix and an improvement.

* Updated the base code from 1.9.0 to 1.16.0.
* Fixed an issue with Minecraft Bedrock 1.16.0: Water in a swampland
  was sometimes not detected and was rendered in the vanilla way.
* Added a configuration item ``ENABLE_FANCY_WATER`` which is enabled
  by default. Now you can leave water unmodified by the shader if you
  wish.

## 1.7.0 -- 2019-02-20

* Updated the base code from 1.8.0 to 1.9.0.
* The shader now abuses ambient occlusion factor to create more
  shadows, especially on the sides of blocks. It can be disabled via a
  configuration item ``ENABLE_OCCLUSION_SHADOWS``.
* Fixed the direction of reflected sunlight which wasn't parallel to
  the sun.
* Added a thin fog that always affects the scene, not only when it's
  raining. Its color and density is currently constant although it
  should ideally be biome-specific (#9). It can be disabled via a
  configuration item ``ENABLE_BASE_FOG``.
* Grass, fern, and vines now have a waving animation (#11). Other
  plants still don't wave though.
* Fixed the configuration item ``ENABLE_WAVES`` not working properly.
* Added a configuration item ``ENABLE_SPECULAR`` so you can disable
  specular lighting which might be laggy.
* Added a configuration item ``ENABLE_RIPPLES`` so you can disable a
  water ripple animation that appears on the ground when it rains.

## 1.6.0 -- 2019-01-30

* Reworked the water again. Now it has a darker base color and shows
  reflected sunlight and moonlight. Note that the direction of
  sunlight doesn't always match to the actual position of the sun due
  to a limitation in the current rendering engine.
* When it's raining the ground now looks wet (#49).
* Changed the color of rain drops so they look more whitish.

## 1.5.0 -- 2019-01-21

* Reworked the water completely. Now it has much more convincing waves
  and specular highlights.
* Clouds now have highlights and shade.
* Shadows no longer disappear all of a sudden when it starts raining
  (#40).

## 1.4.0 -- 2019-01-14

* Fixed vanilla clouds being shown when "Fancy Graphics" is disabled.
* The scene no longer becomes unplayably dark when underwater or in
  The End.
* Moonlight no longer affects the scene on a rainy night.
* The color of the ambient light is no longer a constant white. It now
  changes depending on the terrain-dependent sunlight level and the
  time-dependent daylight level. That is, when in an occluded area
  like a cave, the only possible light sources are torches so the
  color of ambient light resembles that of torch light. Likewise, when
  on the ground of the Overworld, either the color of the sun or the
  moon will affect the ambient depending on the in-game time.
* Changed the fog type from linear fog (vanilla) to exponential
  squared fog (#12). It is slightly more expensive but produces better
  results.
* Distant terrain is now rendered with gradually reduced contrast to
  express a light scattering effect caused by particles flowing in the
  air (#5).
* Cluster of torches no longer flicker in a synchronized manner (#48).

## 1.3.0 -- 2019-01-06

* Clouds are now slightly brighter at night.
* Introduced a randomness in the brightness of stars.
* Torches now flicker more randomly and slightly more intensely.
* Overhauled the entire lighting passes in the terrain shader. Now it
  computes light levels and colors completely differently. This was
  inevitable to improve lights and fix bugs listed below:

    * The intensity of the ambient light now slightly increases when
      it's foggy (#32).
    * The scene no longer becomes reddish as if it were in the sunset,
      when it rains during the daytime (#24).
    * Dark areas, especially caves, no longer become strangely
      brighter when it rains.
    * Shadows are now much less dense at night.

## 1.2.0 -- 2018-12-31

Enhancements:

* Added fake specular light for water and improved its wave. The
  opacity of water surface now changes depending on the angle between
  the camera and the surface. There is still a room for improvement in
  the specular light though. (#10)

Bugfixes:

* Lava pools in the Nether no longer look strangely whitish. (#30)
* Performance improvement for clouds.
* Fixed the color and the intensity of ambient light being calculated
  incorrectly.
* Shadows no longer disappear completely when it rains. It still fades
  suddenly and we're looking for a way to fix it. (#33)
* The animation speed of leaves no longer accelerate when the camera
  is moving. Water still has the same issue but we don't know how to
  fix it. (#36)

## 1.1.0 -- 2018-12-26

Enhancements:

* The scene now looks duller under a bad weather condition. (#6)

src/shaders/glsl/natural-mystic-utils.h
// -*- glsl -*-
#if !defined(NATURAL_MYSTIC_UTILS_H_INCLUDED)
#define NATURAL_MYSTIC_UTILS_H_INCLUDED 1

#include "natural-mystic-config.h"

/* Calculate the luma of a color in the linear RGB color space. */
float rgb2luma(vec3 color) {
    return dot(color, vec3(0.22, 0.707, 0.071));
}

/* Desaturate a color in the linear RGB color space. The parameter
 * "degree" should be in [0,1] where 0 being no desaturation and 1
 * being full desaturation (completely gray). Note that the result of
 * the function is usually to be multiplied by the color of the
 * ambient light, or otherwise a violation of the law of conservation
 * of energy will happen (#30).
 */
vec3 desaturate(vec3 baseColor, float degree) {
    float luma = rgb2luma(baseColor);
    return mix(baseColor, vec3(luma), degree);
}

/* Convert linear RGB to HSV. The x component of the result will be
 * the hue, y will be the saturation, and z will be the value. It does
 * not change the alpha. */
vec4 rgb2hsv(vec4 rgb) {
    highp float rgbMax = max(rgb.r, max(rgb.g, rgb.b));
    highp float rgbMin = min(rgb.r, min(rgb.g, rgb.b));
    highp float h      = 0.0;
    highp float s      = 0.0;
    highp float v      = rgbMax;
    highp float delta  = rgbMax - rgbMin;

    if (delta != 0.0) {
        if (rgbMax == rgb.r) {
            // Between yellow and magenta.
            h = (rgb.g - rgb.b) / delta;
        }
        else if (rgbMax == rgb.g) {
            // Between cyan and yellow;
            h = 2.0 + (rgb.b - rgb.r) / delta;
        }
        else {
            // Between magenta and syan.
            h = 4.0 + (rgb.r - rgb.g) / delta;
        }
    }
    h *= 60.0; // degree
    h  = h < 0.0 ? h + 360.0 : h;

    if (rgbMax != 0.0) {
        s = delta / rgbMax;
    }

    return vec4(h, s, v, rgb.a);
}

/* Generate a random scalar based on some 2D vector. See
 * https://thebookofshaders.com/13/
 */
highp float random(highp vec2 st) {
    st += 128.0; // The function has a bad characteristic near (0, 0).
    return fract(
        sin(dot(st.xy, vec2(12.9898, 78.233))) * 43758.5453123);
}

/* Generate a 2D perlin noise based on some 2D vector. Based on Morgan
 * McGuire @morgan3d https://www.shadertoy.com/view/4dS3Wd
 */
highp float perlinNoise(highp vec2 st) {
    highp vec2 i = floor(st);
    highp vec2 f = fract(st);

    // Four corners in 2D of a tile.
    highp float a = random(i);
    highp float b = random(i + vec2(1.0, 0.0));
    highp float c = random(i + vec2(0.0, 1.0));
    highp float d = random(i + vec2(1.0, 1.0));

    highp vec2 u = f * f * (3.0 - 2.0 * f);

    return mix(a, b, u.x) +
        (c - a) * u.y * (1.0 - u.x) +
        (d - b) * u.x * u.y;
}

/* Generate an fBM noise based on some 2D vector. See
 * https://thebookofshaders.com/13/
 */
highp float fBM(int octaves, highp vec2 st) {
    // Initial values
    highp float value = 0.0;
    highp float amplitude = 0.5;

    // Loop of octaves
    for (int i = 0; i < octaves; i++) {
        value += amplitude * perlinNoise(st);
        st *= 2.0;
        amplitude *= 0.5;
    }
    return value;
}

/* Calculate the color of the ambient light based on some color,
 * usually the fog color, by normalizing the RGB components so at
 * least one component becomes 1.0.
 */
vec3 ambientColor(vec3 baseColor) {
    float rgbMax = max(baseColor.r, max(baseColor.g, baseColor.b));
    float delta  = 1.0 - rgbMax;
    return baseColor + delta;
}

/* Calculate the color and the intensity of the ambient light, based
 * on the fog color. */
vec4 ambientLight(vec4 fogColor) {
    /* THINKME: The existence of fog should increase the intensity of
     * ambient light (#32). */
    return vec4(
        mix(vec3(1.0), ambientColor(fogColor.rgb), fogColor.a),
        0.1);
}

/* Apply the ambient light on the original fragment "frag". The .a
 * component denotes the intensity. Without this filter, objects
 * getting no light will be rendered in complete darkness, which isn't
 * how the reality works.
 */
vec3 applyAmbientLight(vec3 frag, vec3 pigment, vec4 ambient) {
    return frag + ambient.a * pigment * ambient.rgb;
}

/* Calculate and apply a shadow on the original fragment "frag". The
 * argument "torchLevel" should be the torch light level [0, 1],
 * "sunLevel" should be the terrain-dependent sunlight level [0, 1],
 * and "daylight" should be the time-dependent daylight level.
 */
vec3 applyShadow(vec3 frag, float torchLevel, float sunLevel, float daylight, float baseDensity) {
    const vec3 shadowColor = vec3(0.0);
    const float minShadow = 0.0;  // [0, 1]
    const float maxShadow = 0.45; // [0, 1]
    const float blur = 0.003; // The higher the more blur.
    const float torchLightCutOff = 0.1; // [0, 1]

    /* The less the sunlight level is, the darker the fragment will
     * be. */
    float density = mix(maxShadow, minShadow,
                        smoothstep(
                            0.870 - blur,
                            0.870 + blur,
                            sunLevel));

    if (density > 0.0) {
        /* The existence of torch light should negate the effect of
         * shadows. The higher the torch light level is, the less the
         * shadow affects the color. However, we also don't want
         * torches to completely drive away the shadow, since the
         * intensity of the torch light wouldn't be comparable to that
         * of sunlight.
         */
        density *= mix(1.0, torchLightCutOff,
                       smoothstep(0.5, 1.0, torchLevel));

        /* NOTE: Lack of daylight should also negate the effect of
         * shadows, because sadly the light map passed by the upstream
         * doesn't take torches into account. But we can't, because if
         * we reduce "density" depending on "daylight", caves will get
         * brighter at night. We could possibly overcome the issue by
         * completely disabling the vanilla lighting (i.e. diffuse *=
         * texture2D(TEXTURE_1, uv1)) and replacing it with an HDR
         * lighting.
         */
        return mix(frag, shadowColor, baseDensity * density);
    }
    else {
        return frag;
    }
}

/* Calculate the torch light flickering factor based on the in-game
 * time.
 */
float torchLightFlicker(highp float time) {
    /* The flicker is the sum of several sine waves with varying
     * frequently, phase, and amplitude. The reason for the final
     * "/ 10.0" is to avoid underflows.
     *
     * Invariant: -1 <= flicker <= 1 (or Bad Things will happen)
     */
    highp float flicker =
        ( sin(time * 11.0      ) * 0.35 + // fast wave
          sin(time *  3.0 + 0.3) * 0.7    // slow wave
        ) / 10.0;

    /* Workaround for MCPE-39749: the uniform TIME might not be a
     * number. See https://bugs.mojang.com/browse/MCPE-39749
     */
    flicker = clamp(flicker, -1.0, 1.0);
    flicker = isnan(flicker) || isinf(flicker) ? 0.0 : flicker;
    return flicker;
}

/* Calculate and apply the torch color on the original fragment
 * "frag". The argument "torchLevel" should be the torch light level
 * [0, 1], "sunLevel" should be the terrain-dependent sunlight level
 * [0, 1], and "daylight" should be the time-dependent daylight
 * level. The "time" is the in-game time, used for the flickering
 * effect.
 */
vec3 applyTorchColor(vec3 frag, float torchLevel, float sunLevel, float daylight, highp float time) {
    const vec3 torchColor = vec3(0.8, 0.3, -0.2);
    const float torchDecay = 0.55; // [0, 1]
    const float baseIntensity = 1.0; // [0, 1]
    const float sunlightCutOff = 0.1; // [0, 1]

    /* The sunlight should prevent torches from affecting the color,
     * but we also have to take the daylight level into account.
     */
    float intensity = max(0.0, torchLevel - torchDecay) * baseIntensity;
    if (intensity > 0.0) {
        intensity *= mix(1.0, sunlightCutOff, smoothstep(0.65, 0.875, sunLevel * daylight));
#if defined(ENABLE_TORCH_FLICKER)
        intensity *= torchLightFlicker(time) + 1.0;
#endif
        return frag + torchColor * intensity;
    }
    else {
        return frag;
    }
}

/* Apply the sunlight on a fragment "frag" based on the time-dependent
 * daylight level "daylight" [0,1]. The argument "sunLevel" should be
 * the terrain-dependent sunlight level [0,1]. The sunlight is
 * yellow-ish red.
 */
vec3 applySunlight(vec3 frag, float sunLevel, float daylight) {
    const vec3 sunColor = vec3(0.7, 0.3, 0.0);

    float intensity = (0.5 - abs(0.5 - daylight)) * sunLevel;
    if (intensity > 0.0) {
        float sunset = dot(frag, vec3(1.0)) / 1.5;
        return mix(frag, sunset * sunColor, intensity);
    }
    else {
        return frag;
    }
}

/* Apply the skylight on a fragment "frag" based on the time-dependent
 * daylight level "daylight" [0,1]. The argument "sunLevel" should be
 * the terrain-dependent sunlight level [0,1]. The skylight is
 * blue-ish white.
 */
vec3 applySkylight(vec3 frag, float sunLevel, float daylight) {
    const vec3 skyColor = vec3(0.8, 0.8, 1.0);
    const float skyLevel = 0.07;

    float intensity = sunLevel * skyLevel * daylight;

    return frag + skyColor * intensity;
}

/* Apply the moonlight on a fragment "frag" based on the
 * time-dependent daylight level "daylight" [0,1]. The argument
 * "torchLevel" should be the torch light level [0,1], and the
 * argument "sunLevel" should be the terrain-dependent sunlight level
 * [0,1]. The moonlight is purple-ish white.
 */
vec3 applyMoonlight(vec3 frag, float torchLevel, float sunLevel, float daylight) {
    const vec3 moonColor = vec3(0.5, 0.8, 0.95);
    const float moonLevel = 0.7;

    /* The reason why we take into account the torch light level is
     * that the intensity of the torch light is far higher than that
     * of moonlight.
     */
    float factor = sunLevel * (1.0 - torchLevel) * moonLevel * (1.0 - daylight);
    if (factor > 0.0) {
        return mix(frag, desaturate(frag, 0.8) * moonColor, factor);
    }
    else {
        return frag;
    }
}

/* Compute the fog color based on a base fog color, and a camera
 * distance. The resulting color should be mixed with the light color
 * using the alpha of the result. This function produces better fogs
 * than those of vanilla (#12). [Currently unused]
 */
vec4 computeFogColor(vec3 baseFog, float dist) {
    // See: http://in2gpu.com/2014/07/22/create-fog-shader/
    const float density = 0.01;

    float fogFactor = 1.0 / exp(pow(dist * density, 2.0));
    fogFactor = clamp(fogFactor, 0.0, 1.0);

    return vec4(baseFog, 1.0 - fogFactor);
}

/* Apply Uncharted 2 tone mapping to the original fragment "frag".
 * See: http://filmicworlds.com/blog/filmic-tonemapping-operators/
 */
vec3 uncharted2ToneMap_(vec3 x) {
    const float A = 0.15;
    const float B = 0.50;
    const float C = 0.10;
    const float D = 0.20;
    const float E = 0.02;
    const float F = 0.30;

    return ((x * (A * x + C * B) + D * E) / (x * (A * x + B) + D * F)) - E / F;
}
vec3 uncharted2ToneMap(vec3 frag, float exposureBias) {
    const float W = 11.2;

    vec3 curr = uncharted2ToneMap_(exposureBias * frag);
    vec3 whiteScale = 1.0 / uncharted2ToneMap_(vec3(W, W, W));
    vec3 color = curr * whiteScale;

    return color;
}

/* Apply ACES filmic tone mapping to the original fragment "x".
 * See:
 * https://knarkowicz.wordpress.com/2016/01/06/aces-filmic-tone-mapping-curve/
 */
vec3 acesFilmicToneMap(vec3 x) {
    const float a = 2.51;
    const float b = 0.03;
    const float c = 2.43;
    const float d = 0.59;
    const float e = 0.14;

    return clamp((x * (a * x + b)) / (x * (c * x + d) + e), 0.0, 1.0);
}

/* Apply a contrast filter to the original fragment "frag". The
 * contrast must be a non-negative number.
 */
vec3 contrastFilter(vec3 frag, float contrast) {
    return (frag - 0.5) * max(contrast, 0.0) + 0.5;
}

/* Apply an HDR exposure filter to the original LDR fragment
 * "frag". The resulting image will be HDR, and need to be tone-mapped
 * back to LDR at the last stage. */
vec3 hdrExposure(vec3 frag, float overExposure, float underExposure) {
    vec3 overExposed   = frag / overExposure;
    vec3 normalExposed = frag;
    vec3 underExposed = frag * underExposure;

    return mix(overExposed, underExposed, normalExposed);
}

#endif /* NATURAL_MYSTIC_UTILS_H_INCLUDED */
